### PR TITLE
fuzz: add verification to xds fuzzer

### PR DESCRIPTION
**Pull request by samflattery**
Commit Message: Add verification to xDS fuzzer
Additional Description: 

- ensures that all listeners are in the correct state
- I have not figured out how to verify draining listeners, since `XdsFuzzTest` inherits from `HttpIntegrationTest` which uses real time, so when i try to inherit from `Event::TestUsingSimulatedTime` I get an assertion error as you can't have 2 different time systems, is `DrainManager` useful in this case?
- it fails on a few of the corpus entries right now because of https://github.com/envoyproxy/envoy/issues/11871 as the active stat on the server is incremented when the warming stat should be and I'm not sure why. 
- this branch is also branched off from `xds_fuzz`, which is the branch I made the original PR with, so the commit history is a bit mangled if I go to make an actual PR to envoy with this.  Is this ok or should I just copy all the code to a fresh branch and commit from there?
- this is still pretty rough so there's some cleaning up I want to do before I make an actual PR but I just wanted to get some feedback that I was going in the right direction with this and any suggestions to improve things

Risk Level: Low
Testing: Ran with libfuzzer for a few minutes, 91% coverage over `source/common/init`
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

**Comment by samflattery** [comment]
I might split this up until multiple PRs and push all of the new waiting for ACKs and stats first since I'm just seeing that the original fuzzer is timing out waiting for messages on OSS fuzz a lot (https://oss-fuzz.com/testcases?fuzzer=libFuzzer_envoy_xds_fuzz_test)

**Comment by asraa** [comment]
Hmmm couple of comments --
* The time issue. I agree SimulatedTime is probably what you want. Does the example in `HttpTimeoutIntegrationTest` help? I saw a docstring somewhere about it that you have to specify it as the first base class.
* Commit history: is it possible to merge HEAD now that the base fuzzer is committed and resolve conflicts so that the only diff that shows is the verification? You can feel free to force push before you make the actual PR (just don't force push after the main envoy PR is made) EDIT: ah! nice!!! you already did that (just checked the code). Yeah, you can force push/squash. That's not an issue. 
* To resolve the waits -- will simulatedtime help if you can get that working?

File: test/server/config_validation/xds_fuzz.cc [hunk 1]
```diff
@@ -184,77 +186,244 @@ void XdsFuzzTest::replay() {
...
+            found = true;
+          }
+          break;
+        default:
+          NOT_REACHED_GCOVR_EXCL_LINE;
+        }
+      }
+    }
+    if (!found) {
+      ENVOY_LOG_MISC(info, "Expected to find listener {} in config dump",
```
Comment (asraa) [inline test/server/config_validation/xds_fuzz.cc:412]: I see. So at a high-level, we keep track of listenres in the verifier, and then go through the list of listeners in the listener dump, find it's representative in the verifier, and compare their states? 'found' is a representation of "is that listener in the verifier AND does it's state match?"

File: test/server/config_validation/xds_fuzz.cc [hunk 2]
```diff
@@ -184,77 +186,244 @@ void XdsFuzzTest::replay() {
...
+  /* timeSystem().advanceTimeWait(std::chrono::milliseconds(300000)); */
+  /* EXPECT_EQ(test_server_->gauge("listener_manager.total_listeners_draining")->value(), */
+  /*           verifier_.numDraining()); */
+  /* timeSystem().advanceTimeWait(std::chrono::milliseconds(300001)); */
+  /* EXPECT_EQ(test_server_->gauge("listener_manager.total_listeners_draining")->value(), */
+  /*           verifier_.numDraining() - 1); */
+  verifier_.drainedListener(name);
+}
+
+void XdsFuzzTest::verifyListeners() {
```
Comment (asraa) [inline test/server/config_validation/xds_fuzz.cc:372]: what do you think about factoring a lot of this logic into the verifier? 

as in, we pass in the listener rep's name and state to the verifier, and the verifier holds the logic in here that compares?

**Comment by asraa** [comment]
I took a really quick first pass -- I will take a more in depth one in the morning, and let's chat at our 1/1 about it!

File: test/server/config_validation/xds_fuzz.cc (hunk 1 above)
Comment (samflattery) [reply test/server/config_validation/xds_fuzz.cc:412]: Right.  For each listener in the verifier, we check to see that there is a matching listener in the config dump, i.e. listener with same name and state

File: test/server/config_validation/xds_fuzz.cc (hunk 2 above)
Comment (samflattery) [reply test/server/config_validation/xds_fuzz.cc:372]: I was thinking of doing this, but then I would need to pass `test_server` in to the verifier class as well to be able to check stats, etc, and `test_server_` is a unique_ptr AFAIK so I'm not sure if I'm able to do this?

**Comment by samflattery** [comment]
> The time issue. I agree SimulatedTime is probably what you want. Does the example in `HttpTimeoutIntegrationTest` help? I saw a docstring somewhere about it that you have to specify it as the first base class.

Yep, inheriting from that class before `HttpIntegrationTest` seemed to do the trick, thanks

> To resolve the waits -- will simulatedtime help if you can get that working?

I'm not too sure, because it seems like the waits i have are legitimate waits, like waiting for ACKs and stats on the server to line up so I'm not sure if simulated time would help here, what do you think?

**Pull request closed without merging**